Add extrema_filter

codles opened the pull request:
Added function and tests for 2d and 4d case

codles commented:
Initial add following [METADATA](https://github.com/JuliaLang/METADATA.jl/pull/4036) discussion.

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+                @nexprs $(N-1) j->(sa_{j} = i_j)
+
+                # Filter the last dimension
+                (@nref $N minval_temp sa) = min_filter(vec( @nref $N minval_temp fa), window[dim])
```
This doesn't seem to be defined (likewise `max_filter`)

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+# Min max filter
+
+type Wedge
+    buffer::AbstractArray
```
For good performance, you're going to want to declare this to be a leaf type. If you need to handle arbitrary array types, then something like
```jl
type Wedge{A<:AbstractArray}
    buffer::A
    size::Int
    ...
end
```
will give you much better performance.

timholy replied:
(See http://docs.julialang.org/en/stable/manual/faq/#how-do-abstract-or-ambiguous-fields-in-types-interact-with-the-compiler)

codles replied:
Thanks

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+
+    for i = 2:n
+        if i > window
+            if ~wedgeisempty(U)
```
! rather than ~? (Matlab-ism?)

codles replied:
Thanks. Old habits die hard

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+end
+
+function pushback(X::Wedge, v)
+    X.last = mod(X.last, X.size) + 1
```
AFAICT, this is just a circular buffer and loops 1,2,3,...,size,1,2,3,...size,.... It turns out that `mod` is slow (it relies on integer division, which is one of the slowest operations on a CPU), so you could probably speed this up with
```jl
X.last += 1
if X.last > X.size
    X.last = 1
end
```
or something.

codles replied:
Again I was unaware of this. Thanks

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+    X.last = mod(X.last, X.size) + 1
+    X.buffer[X.last] = v
+    X.n = X.n+1
+    X.mxn = maximum([X.mxn, X.n])
```
`max(X.mxn, X.n)` will be orders-of-magnitude faster.

codles replied:
I didn't know this. Thanks

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+    X.n <= 0
+end
+
+function pushback(X::Wedge, v)
```
Maybe call this `pushback!`? Functions which mutate their inputs conventionally end in `!`.

codles replied:
Noted, thanks

codles commented:
Thanks @timholy. Ill make your suggested changes. And a few changes of my own.
Should I build on this pull request? Or make the changes and open a new pull request?

timholy commented on src/algorithms.jl:
```diff
@@ -1473,6 +1473,171 @@ extr(order::Ordering, x::Color, y::Color) = extr(order, convert(RGB, x), convert
+                @nexprs $(N)   j->(sa_{j} = 1:size(maxval_temp)[j] - window[dim] + 1)
+                @nexprs $(N-1) j->(sa_{j} = i_j)
+
+                # Filter the last dimension
```
It seems that you're actually filtering on the first dimension? Julia arrays are column-major.

timholy commented:
You can just build on this one. When you push new commits to `codles:master`, this will be automatically updated.

I don't get an email each time you push a commit, so when you're ready for me to check this again just post a comment here.

codles commented:
@timholy I think I have addressed your concerns above and added tests for all lines of code. Please let me know of any additional changes you want made.

timholy commented:
Yay, thanks!
